cmd/source_profile.go

```diff
@@ -0,0 +1,274 @@
+	awsSecretAccessKey string // Same as AWS_SECRET_ACCESS_KEY environment variable
+	awsRegion          string // Same as AWS_REGION environment variable
+	dydbEndpoint       string // Same as DYNAMODB_ENDPOINT_OVERRIDE environment variable
+	schemaSampleSize   int64  // Number of rows to use for inferring schema (default 100,000)
+}
+
+func NewSourceProfileConnectionDynamoDB(params map[string]string) (SourceProfileConnectionDynamoDB, error) {
+	// TODO: Populate environment variables for DynamoDB if passed as flags.
+	dydb := SourceProfileConnectionDynamoDB{}
+	if awsAccessKeyID, ok := params["awsAccessKeyID"]; ok {
+		dydb.awsAccessKeyID = awsAccessKeyID
+	}
+	if awsSecretAccessKey, ok := params["awsSecretAccessKey"]; ok {
+		dydb.awsSecretAccessKey = awsSecretAccessKey
+	}
+	if awsRegion, ok := params["awsRegion"]; ok {
+		dydb.awsRegion = awsRegion
+	}
+	if dydbEndpoint, ok := params["dydbEndpoint"]; ok {
+		dydb.dydbEndpoint = dydbEndpoint
+	}
+	if schemaSampleSize, ok := params["schemaSampleSize"]; ok {
+		schemaSampleSizeInt, err := strconv.Atoi(schemaSampleSize)
+		if err != nil {
+			return dydb, fmt.Errorf("could not parse schemaSampleSize = %v as a valid int64", schemaSampleSize)
+		}
+		dydb.schemaSampleSize = int64(schemaSampleSizeInt)
+	}
+	return dydb, nil
+}
+
+type SourceProfileConnection struct {
+	ty    SourceProfileConnectionType
+	mysql SourceProfileConnectionMySQL
+	pg    SourceProfileConnectionPostgreSQL
+	dydb  SourceProfileConnectionDynamoDB
+}
+
+func NewSourceProfileConnection(source string, params map[string]string) (SourceProfileConnection, error) {
+	conn := SourceProfileConnection{}
+	switch strings.ToLower(source) {
+	case "mysql":
+		{
+			conn.ty = SourceProfileConnectionTypeMySQL
+			conn.mysql = NewSourceProfileConnectionMySQL(params)
+		}
+	case "postgresql", "postgres", "pg":
+		{
+			conn.ty = SourceProfileConnectionTypePostgreSQL
+			conn.pg = NewSourceProfileConnectionPostgreSQL(params)
+		}
+	case "dynamodb":
+		{
+			conn.ty = SourceProfileConnectionTypeDynamoDB
+			dydb, err := NewSourceProfileConnectionDynamoDB(params)
+			conn.dydb = dydb
+			if err != nil {
+				return conn, err
+			}
+		}
+	default:
+		return conn, fmt.Errorf("invalid source database: %v", source)
+	}
+	return conn, nil
+}
+
+type SourceProfileConfig struct {
+	path string
+}
+
+func NewSourceProfileConfig(path string) SourceProfileConfig {
+	return SourceProfileConfig{path: path}
+}
+
+type SourceProfile struct {
+	ty     SourceProfileType
+	file   SourceProfileFile
+	conn   SourceProfileConnection
+	config SourceProfileConfig
+}
+
+func (src SourceProfile) ToLegacyDriver(source string) (string, error) {
+	switch src.ty {
+	case SourceProfileTypeFile:
+		{
+			switch strings.ToLower(source) {
+			case "mysql":
+				return "mysqldump", nil
+			case "postgresql", "postgres", "pg":
+				return "pg_dump", nil
+			default:
+				return "", fmt.Errorf("invalid source database %v for source-profile type file", strings.ToLower(source))
```

Comment:
Suggest rephrasing as:

invalid source=%s specification in source profile

I think that makes it easier for users to track down the issue (and "source database" is kind of ambiguous here, since it could refer to the source database type or the source database instance).

Also, I'm not sure users will know what "source-profile type file" means?
I realize you want to say what mode we are running in, since e.g. DynamoDB isn't supported with a file input. I'm wondering whether it would be clearer to separate out the DynamoDB case and generate a targeted message in this case like:
"Dump files are not supported with DynamoDB."

Reply:
`source=` isn't part of the source profile but is rather is a separate flag (similarly for `target =`). I agree with catching DynamoDB case separately and have implemented the same. I have also improved the overall error messaging.

Reply:
I have added top-level error checking for source database which checks for empty source - "cannot leave -source flag empty, please specify source databases e.g., -source=postgres etc"; more specific checks are done separately where context is available.